src/ps.c

```diff
@@ -0,0 +1,336 @@
+#include "bouncer.h"
+
+#include <usual/crypto/csrandom.h>
+#include <usual/hashing/spooky.h>
+#include <usual/hashtab-impl.h>
+
+static PgParsedPreparedStatement *create_prepared_statement(PgParsePacket *parse_packet)
+{
+  static bool initialized;
+	static uint32_t rand_seed;
+  PgParsedPreparedStatement *s = NULL;
+
+	if (!initialized) {
+		initialized = true;
+		rand_seed = csrandom();
+	}
+
+  s = (PgParsedPreparedStatement *)malloc(sizeof(*s));
+  s->name = parse_packet->name;
+  s->pkt = parse_packet;
+  s->query_hash[0] = rand_seed;
+	s->query_hash[1] = 0;
+	spookyhash(parse_packet->query, strlen(parse_packet->query), &s->query_hash[0], &s->query_hash[1]);
+  
+  return s;
+}
+
+static PgServerPreparedStatement *create_server_prepared_statement(PgSocket *client, PgParsedPreparedStatement *ps)
+{
+  PgServerPreparedStatement *s = NULL;
+  char statement[23];
+
+  s = (PgServerPreparedStatement *)malloc(sizeof(*s));
+  *(uint64_t *)&s->query_hash[0] = ps->query_hash[0];
+  *(uint64_t *)&s->query_hash[1] = ps->query_hash[1];
+  s->bind_count = 0;
+
+  // FIXME: length??
+  snprintf(statement, 23, "B_%ld", client->link->nextUniquePreparedStatementID++);
+  s->name = strdup(statement);
+
+  return s;
+}
+
+static int bind_count_sort(struct PgServerPreparedStatement *a, struct PgServerPreparedStatement *b) {
+    return (a->bind_count - b->bind_count);
+}
+
+static bool register_prepared_statement(PgSocket *server, PgServerPreparedStatement *stmt)
+{
+  struct PgServerPreparedStatement *current, *tmp;
+  PktBuf *buf;
+  unsigned int cached_query_count = HASH_COUNT(server->server_prepared_statements);
+
+  if (cached_query_count >= (unsigned int)cf_prepared_statement_cache_queries)
+  {
+    HASH_SORT(server->server_prepared_statements, bind_count_sort);
```

Comment:
Such replacement algorithm will cause throwing away newly added statements just because their bond_count  is small.
It will be better to use Clock or LRU replacement algorithm.
Also it will help to eliminate sort overhead on each new item which may become bottleneck for large number of prepared statements.